src/DiffResults.jl

```diff
@@ -36,7 +36,7 @@ Return `r::DiffResult`, with output value storage provided by `value` and output
 storage provided by `derivs`.
 
 In reality, `DiffResult` is an abstract supertype of two concrete types, `MutableDiffResult`
-and `ImmutableDiffResult`. If all `value`/`derivs` are all `Number`s or `StaticArray`s,
+and `ImmutableDiffResult`. If all `value`/`derivs` are all `Number`s or `StaticArrays.SArray`s,
```

Comment:
That way, if there is even one mutable array in the mix (like `MArray`), the whole thing becomes a `MutableDiffResult`

Reply:
you're right, so it's the other way around

Reply:
Yes, to me it seems that's what we want for StaticArrays. The safest option would be to make the `ImmutableDiffResult` the default `DiffResult` type, but given the wide range of mutable arrays (without clear way to catch them all apart from - to some extent - `ArrayInterface.ismutable`) I think it's much easier implementation-wise to only add special cases for StaticArrays. Even though this is also doomed to fail with wrapper types such as `Diagonal` etc...

Reply:
Theoretically, if everyone used the API correctly, making `Immutable` the default would not be breaking. Unfortunately, as seen in https://github.com/JuliaDiff/DiffResults.jl/issues/26, SciML and others are playing fast and loose with realiasing.

So from an ecosystem point of view, any transition `Mutable` -> `Immutable` has the potential to be breaking